### ISSUE TITLE
Adding timing imprecision

### ISSUE THREAD

**Issue by lud** [beatboxjs/beatbox.js#2]
Hello,

I would like to add a simple feature to my beat : natural imprecision of a human player.

If my pattern is divided in 8 beats, I would not have to transform it to a 800 beats patterns to add small delays or advance for each beat.

Could you give me some informations on how to implement that ?

I was thinking about overriding `Beatbox._playWhen` :

```
static _playWhen(instrumentWithParams, when) {
    when += Math.random() / 100
    …
}
```

But I am not sure of what `_whenOverride` is for and if it should be changed.

Also you seem to override `window.setTimeout` at each call of `_playWhen`. It does not seem possible to override this function in window as `window.setTimeout.toString()` yields:

```
"function setTimeout() {
    [native code]
}"
```
Does it work only in your scope ?

Why do you need to override it ? Could you not simply use a custom function in your module, or is it for Howler to use it ?

Thank you.

**Comment by cdauth** [comment]
Your approach should work.

Howler does not provide a way to specify the `when` parameter. Thus I'm overriding `Howler.ctx.createBufferSource().start()` and manually set the `when` parameter to the value of `Beatbox._whenOverride`, and I'm overriding the `setTimeout` function so that the Howler cleanup runs in the right moment. The fact that it is a native function doesn't affect its ability to be overridden.

**Comment by lud** [comment]
Thank you for your answer !

I am reading all the code and it is very interesting, although a bit complicated as a first approach to web audio :)

Ok I got it backwards, `Beatbox._setTimeout` is actually a reference to the native implemetation. I could not understand why you would override `window.setTimeout` two times. Now I understant that the override is temporary and that is why the toString showed the native implemetation in the console.

It seems that I cannot add more time than the duration of a beat, or the sound will not play. Maybe there is som code that cancels the sounds scheduled too far in the future ?

Also, I found that I have to change the `when` value before assigning `_whenOverride` to it. If I change it after, there is no random delay, as if it changes nothing ; which is strange because the value is used in the `setTimeout` override … So the actual scheduling delay would be done by `arguments[0] = Beatbox._whenOverride` ? And the change to `millis` would be only about the cleanup, that has to be delayed or the sounds get removed before being played … I know I am wrong because if I comment the actual call to `Beatbox._setTimeout` nothing is played.

Well that is hard =)

Thanks for your time !
